* Manuscript and Data

  This repository contains source code developed for this [[https://www.biorxiv.org/content/early/2018/05/04/314815][manuscript]]:

  #+BEGIN_QUOTE
  Selfish mutations dysregulating RAS-MAPK signaling are pervasive in aged human testes

  Geoffrey J. Maher, Hannah K. Ralph, Zhihao Ding, Nils Koelling, Hana
  Mlcochova, Eleni Giannoulatou, Pawan Dhami, Dirk S. Paul, Stefan
  H. Stricker, Stephan Beck, Gilean McVean, Andrew OM Wilkie, Anne
  Goriely
  #+END_QUOTE

  The raw FASTQ files can be downloaded from here.  The VCF with
  pileup counts per sample per site can be downloaded here.

* User guide

  - Scan reads in FASTQ for sample. Assign amplicon ID to each read
    using a kmer search algorithm.

    #+BEGIN_SRC sh
      python parse_primer.py --fastq --sample sampleName
    #+END_SRC

  - Alignment for each sample.

    #+BEGIN_SRC sh
      bwa mem -R "@RG\tID:$ID\tSM:$SM\tLB:$LB" $ref $read1 $read2 > $sam
    #+END_SRC

  - Pileup for all sites for each sample.

    #+BEGIN_SRC sh
      python parse_primer.py --pileup --sample sampleName
    #+END_SRC

  - Combine pileups of all samples and write out result datasets.

    #+BEGIN_SRC sh
      python parse_primer.py --sum
    #+END_SRC

  - Generating files/plots for each amplicon to assist QC on input
    signals.

    #+BEGIN_SRC sh
      python parse_primer.py --agg --amp ampliconName
    #+END_SRC

  - Call elevated alleles for amplicon.

    #+BEGIN_SRC sh
      python parse_primer.py --call --amp ampliconName
    #+END_SRC


* Meta Data

  | File Name           | Description                               |
  |---------------------+-------------------------------------------|
  | meta/samplelist     | A list of IDs of all samples              |
  | meta/fastq.index    | A list input  fastq files                 |
  | meta/amplicons.csv  | Meta information for the  amplicons       |
  | meta/sample_seq.csv | Meta information for sequencing libraries |
