# MIP

MIP related analysis.

This is a trivial demo for some simple analyses
parsing reads generated by MIP capture followed by
sequencing. 


### Aim

- For each read, align the Illumina common sequence against it to find
  out where the common sequence ends. This may be tricky as we need to
  tweak alignment parameters.

- Quantify the diversity of the random pentamers inserted
  experimentally that follow immediately the common sequence in each
  read against one or more variants captured by the read, such that we
  know how many independent molecules supporting the variant.

### Run

```
python parse.py
```

### Parse MiSeq data

- We want to quantify the number of molecular identifiers for each
  input file. 

### Log printed to screen

```
INFO:root:loaded 1 probes
INFO:root:loaded 48 barcodes
INFO:root:BAM:BC1_rawlib.bam
INFO:root: tag:BC1
ATCGGGAAGCTGAAGC-A-AGAGCAGACAAACTCTACGTCTCCTCCGACCAC
--------------GCTATAG-GCTGCCA---TC-ACA--------------
14,37,23
CAAACTCTACGTCTCCTCCGACCACTGTGGAGGCATTTGCCGGCAGTC-CG
-------------------------GCTATAGGC---TGCC---A-TCACA
30,50,20
ACTCTACCGTCTCCTCCGACCACTGTGGAGGCATTTGCCGGCAGTC-CGGT
-----------------------GCTATAGGC---TGCC---A-TCACA--
28,48,20
CCGACACTGTGGAGGCATTTGCCGGCAGTC-CGGTTGGAGGATGGCCGGTG
-------GCTATAGGC---TGCC---A-TCACA------------------
12,32,20
CGATATCCGACGGTAGTGTAAGTTTGGAGGATGGGCCGGTGAGGCGATCG
CGATATCCGACGGTAGTGT-------------------------------
0,19,19
CGATATCCGACGGTAGTGTCGAAGTGGAGGATGGGCCGGTGAGGCGATCG
CGATATCCGACGGTAGTGT-------------------------------
0,19,19
CGATATCCGACGGTAGTGTGTACATGGAGGATGGGCCGGTGAGGCGATCG
CGATATCCGACGGTAGTGT-------------------------------
0,19,19
ATTTGCCGGCAGTCCGGTTTGGGAGGGATGGGCCGGTGAGGC-G--ATCGCTC
--------------------------------GCTAT-AGGCTGCCATCACA-
36,51,15
GCCGGCAGTCCGGCT-TGGAGGATGGCCGGTGAGGCGATCGCTCTGGTGGA
------------GCTAT--AGGCTG-CCATCACA-----------------
12,28,16
INFO:root:processed 10000 reads
CAGTCCGGTTGGAGGATGGGCCGGTGAGGC-G--ATCGCTCTGGTGGAGAGAG
--------------------GCTAT-AGGCTGCCATCACA-------------
24,39,15
GGAGGATGGGCGGTGAGGCGATGC--TCTGTGGAGAGAGGGAAGAAAGGAGG
---------GCTAT-AGGC--TGCCATCACA---------------------
13,28,15
GAGGATGGCCGGTGAGGCGATGC--TCTGGTGGAGAGAGGGAAGAAAGGAGG
--------GCTAT-AGGC--TGCCATCACA----------------------
12,27,15
```

### Output file

```
forward,BC1,A,GGGTG
forward,BC1,A,CTACT
forward,BC1,A,TTGCG
forward,BC1,A,AGTCT
forward,BC1,A,CGTGG
forward,BC1,A,GCGCA
forward,BC1,A,TCGGC
forward,BC1,A,CGCTC
forward,BC1,A,GGTCT
forward,BC1,A,GCACG
....
```
